### PR TITLE
Update RuleBuilder to effectively handle Patient Type-Level Exports

### PR DESCRIPTION
**Pull request by Chris0296**
Issue: https://gitlab.com/simpatico.ai/cdr/-/issues/4634

Rule builder could not effectively handle Patient Type-Level Exports. It would over-permit requests in certain scenarios.
    
This fix allows for accumulation of ids on a Patient Type-Level Bulk export to enable us to properly match the requested Patient IDs against the users permitted Patient IDs.

### CODE REVIEW

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/4634-rulebuilder-bulkexport-patient.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    
+    WHEN they submit a patient bulk export request which specifies these patients to export:
+    Patient/1
+    Patient/2
+    
+    THEN the bulk export request would be accepted, based on the fact that the user has permission for Patient/1, 
+    despite the fact they they did not have permission for Patient/2
+    
+    This fix allows for accumulation of ids on a Patient Type-Level Bulk export to enable us to properly match the
+    requested Patient IDs against the users permitted Patient IDs."
```
Comment (tadgh) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/4634-rulebuilder-bulkexport-patient.yaml:18]: Issue: this changelog is too expansive I think. Its sufficient to say it was over-permitting, and is no longer. The formatting you've put in here likely won't carry over with our renderer either way.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java [hunk 2]
```diff
@@ -147,7 +157,15 @@ public AuthorizationInterceptor.Verdict applyRule(
...
-				if (patientIdsInFilters.contains(expectedPatientId)) {
+				boolean filteredPatientIdsPermitted = true;
+				for (String patientIdInFilters : patientIdsInFilters) {
+					if (!permittedPatientIds.contains(patientIdInFilters)) {
+						filteredPatientIdsPermitted = false;
+						break;
+					}
+				}
+
+				if (filteredPatientIdsPermitted) {
```
Comment (tadgh) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java:168]: ah nice fix

File: hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBuilderTest.java [hunk 3]
```diff
@@ -87,6 +88,19 @@ public void testBulkExportPermitsIfASingleGroupMatches() {
...
+		List<String> resourceTypes = new ArrayList<>();
+		resourceTypes.add("Patient");
+
+		builder.allow().bulkExport().patientExportOnPatient("Patient/pat1").withResourceTypes(resourceTypes);
+		builder.allow().bulkExport().patientExportOnPatient("Patient/pat2").withResourceTypes(resourceTypes);
+		List<IAuthRule> rules = builder.build();
+		assertEquals(rules.size(),1);
+		assertTrue(rules.get(0) instanceof RuleBulkExportImpl);
+	}
+
```
Comment (tadgh) [inline hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBuilderTest.java:103]: Issue: While this test is useful, it is not sufficient to test the failure scenario. As far as I can tell, this just tests that it generates one rule, and that it happens to be the correct type. A more useful test would be to show that attempting to request a bulk export for a specific patient that the user does not have access to gets denied. That would prove that the bug is fixed more conclusively than this test.

**Review by tadgh — APPROVED** [review]
Approved pending some minor comments and addition of another test

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/4634-rulebuilder-bulkexport-patient.yaml (hunk 1 above)
Comment (Chris0296) [reply hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/7_0_0/4634-rulebuilder-bulkexport-patient.yaml]: Removed excess from the changelog

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java (hunk 2 above)
Comment (Chris0296) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImpl.java:164]: Thanks!

File: hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBuilderTest.java (hunk 3 above)
Comment (Chris0296) [reply hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBuilderTest.java:103]: I have added more conclusive tests!

File: hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImplTest.java [hunk 4]
```diff
@@ -249,4 +249,41 @@ public void testPatientExportRulesOnTypeLevelExportWithTypeFilterMismatch() {
...
+		options.setFilters(Set.of("Patient?_id=123","Patient?_id=456"));
+		options.setResourceTypes(Set.of("Patient"));
+		when(myRequestDetails.getAttribute(any())).thenReturn(options);
+
+		//When
+		final AuthorizationInterceptor.Verdict verdict = myRule.applyRule(myOperation, myRequestDetails, null, null, null, myRuleApplier, myFlags, myPointcut);
+
+		//Then: There are unpermitted patients in the request so this is not permitted.
+		assertEquals(PolicyEnum.DENY, verdict.getDecision());
+	}
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImplTest.java:288]: Please add the three cases from the IT you created:
- auth for patient1, request export with patient param of patient2, expect DENY
- auth for patient1, request export with patient=patient1, expect ALLOW
- auth for patient1, request export with patient=patient1,patient2 expect DENY
- auth for patient1 and patient2, request export with patient=patient1,patient2 expect ALLOW

Comment (Chris0296) [reply hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleBulkExportImplTest.java:346]: Done!

**Pull request merged**